src/Tracer.ts

```diff
@@ -1,297 +1,79 @@
-   * @param traceOptions options for status and event action. Default action is 'span' and status is success
-   */
-  async trace(traceContext: TraceContext = this._traceContext, traceOptions: LoggerOptions = { action: TraceEventAction.span }): Promise<any> {
-    let { state } = extractLoggerOptions(EventType.trace, traceOptions)
-    if (this._traceContext.finishTimestamp && this.finished) throw new Error('span already finished')
-    if (!state) throw new Error('no valid state provided')
-    let event = EventMetadata.trace({ action: TraceEventAction.span, state })
-    let message = new EventMessage({
-      type: 'trace',
-      content: traceContext,
-      metadata: {
-        event,
-        trace: <IEventTrace>traceContext
-      }
-    })
-    let logResult = await this.record(message);
-    if (LogResponseStatus.accepted == logResult.status) {
-      this.finished = this._traceContext.finishTimestamp ? true : false
-      return logResult
-    } else {
-      throw new Error(`Error when logging trace. status: ${logResult.status}`)
-    }
+  static createChildSpanFromContext(service: string, spanContext: TypeSpanContext, recorders?: Recorders): Span {
+    let outputContext = <TypeSpanContext>Object.assign({}, spanContext, { service, spanId: undefined, parentSpanId: spanContext.spanId, startTimestamp: undefined })
+    return new Span(new EventTraceMetadata(outputContext), recorders) as Span
   }
 
   /**
-   * Sends audit type message to the event logging framework. 
-   * @param message message to be recorded as audit event
-   * @param auditOptions Logger options object.
+   * Injects trace context into a carrier with optional path.
+   * @param carrier any kind of message or other object with keys of type String.
+   * @param injectOptions type and path of the carrier. Type is not implemented yet. Path is the path to the trace context.
    */
-  async audit(message: EventMessage, auditOptions: LoggerOptions = { action: AuditEventAction.default }): Promise<any> {
-    let { action, state } = extractLoggerOptions(EventType.audit, auditOptions)
-    if (!action) throw new Error('no valid action provied')
-    if (!state) throw new Error('no valid state provided')
-    if (this.finished) throw new Error('span already finished')
-    let newEnvelope = new EventMessage(Object.assign(message, {
-      metadata: {
-        event: EventMetadata.audit({
-          action,
-          state
-        }),
-        trace: this._traceContext
+  static injectContextToMessage(context: TypeSpanContext, carrier: { [key: string]: any }, injectOptions: ContextOptions = {}): Promise<{ [key: string]: any }> {
+    let result = carrier
+    let { path } = injectOptions // type not implemented yet
+    if (carrier instanceof EventMessage || (('metadata' in carrier))) path = 'metadata.trace'
+    else if (('trace' in carrier)) path = 'trace'
+    else if (carrier instanceof EventTraceMetadata) return Promise.resolve(context)
+    if (path) {
```

Comment:
Any reason why you are not using something like lodash to help you with this magic trickery?

I think it would make this more readable, and remove any chance of bugs, etc.